draft-barnes-acme.md

```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
+    "server": { "kty": "EC", ... }
+  }
+}
+
+~~~~~~~~~~
+
+On receiving the server's response, the client can compute the recovery key by combining the server's public key together with the private key corresponding to the public key that it sent to the server.
+
+Clients may refresh the recovery key associated with a registration by sending a POST request with a new recoveryKey object.  If the server agrees to refresh the recovery key, then it responds in the same way as to a new registration request that asks for a recovery key.
+
+~~~~~~~~~~
+
+POST /acme/reg/asdf HTTP/1.1
+Host: example.com
+
+{
+  "resource": "reg",
+  "recoveryKey": {
+    "client": { "kty": "EC", ... }
+  }
+}
+/* Signed as JWS */
+
+~~~~~~~~~~
+
+## Account Recovery
 
 Once a client has created an account with an ACME server, it is possible that the private key for the account will be lost.  The recovery contacts included in the registration allows the client to recover from this situtation, as long as it still has access to these contacts.
 
-{::comment}
-TODO: Re-add recoveryContact here https://github.com/letsencrypt/acme-spec/issues/136
-{:/comment}
+By "recovery", we mean that the information associated with an old account key is bound to a new account key.  When a recovery process succeeds, the server provides the client with a new registration whose contents are the same as base registration object -- except for the "key" field, which is set to the new account key.  The server reassigns resources associated with the base registration to the new registration (e.g., authorizations and certificates).  The server SHOULD delete the old registration resource after it has been used as a base for recovery.
+
+In addition to the recovery mechanisms defined by ACME, individual client implementations may also offer implementation-specific recovery mechanisms.  For example, if a client creates account keys deterministically from a seed value, then this seed could be used to recover the account key by re-generating it.  Or an implementation could escrow an encrypted copy of the account key with a cloud storage provider, and give the encryption key to the user as a recovery value.
+
+### Token-Based Recovery
+
+With token-based recovery, the client proves to the server that it holds a secret value established in the initial registration transaction.  The client requests token-based recovery by sending a MAC over the new account key, using the recovery key from the initial registration.
+
+method (required, string):
+: The string "token"
+
+base (required, string):
+: The URI for the registration to be recovered.
+
+token (required, string):
+: A JSON-formatted JWS object using an HMAC algorithm, whose payload is the JWK representation of the public key of the new account key pair.
+
+~~~~~~~~~~
+
+POST /acme/recover-registration HTTP/1.1
+Host: example.com
+
+{
+  "resource": "recover-reg",
+  "method": "token",
+  "base": "https://example.com/acme/reg/asdf",
```

Comment:
Suggestion?  "authorization"?  Since the MAC asserts the new key under the authority of the recovery key holder.